lib/sqlite/sqlite.go

```diff
@@ -0,0 +1,165 @@
+		return nil, false, xerrors.Errorf("error creating database base directory [@ %s]: %w", path, err)
+	}
+
+	_, err := os.Stat(path)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, false, xerrors.Errorf("error checking file status for database [@ %s]: %w", path, err)
+	}
+	exists := err == nil
+
+	db, err := sql.Open("sqlite3", path+"?mode=rwc")
+	if err != nil {
+		return nil, false, xerrors.Errorf("error opening database [@ %s]: %w", path, err)
+	}
+
+	for _, pragma := range pragmas {
+		if _, err := db.Exec(pragma); err != nil {
+			_ = db.Close()
+			return nil, false, xerrors.Errorf("error setting database pragma %q: %w", pragma, err)
+		}
+	}
+
+	return db, exists, nil
+}
+
+// InitDb initializes the database by checking whether it needs to be created or upgraded.
+// The ddls are the DDL statements to create the tables in the database and their initial required
+// content. The schemaVersion will be set inside the databse if it is newly created. Otherwise, the
+// version is read from the databse and returned. This value should be checked against the expected
+// version to determine if the database needs to be upgraded.
+func InitDb(
+	ctx context.Context,
+	name string,
+	db *sql.DB,
+	ddls []string,
+	versionMigrations []MigrationFunc,
+) error {
+
+	schemaVersion := len(versionMigrations) + 1
+
+	q, err := db.QueryContext(ctx, "SELECT name FROM sqlite_master WHERE type='table' AND name='_meta';")
+	if q != nil {
+		defer func() { _ = q.Close() }()
+	}
+
+	if errors.Is(err, sql.ErrNoRows) || !q.Next() {
+		// empty database, create the schema including the _meta table and its versions
+		ddls := append(metaDdl(uint64(schemaVersion)), ddls...)
+		for _, ddl := range ddls {
+			if _, err := db.Exec(ddl); err != nil {
+				return xerrors.Errorf("failed to %s database execute ddl %q: %w", name, ddl, err)
+			}
+		}
+		return nil
+	}
+
+	if err != nil {
+		return xerrors.Errorf("error looking for %s database _meta table: %w", name, err)
+	}
+
+	if err := q.Close(); err != nil {
+		return xerrors.Errorf("error closing %s database _meta table query: %w", name, err)
+	}
+
+	// check the schema version to see if we need to upgrade the database schema
+	var foundVersion int
+	if err = db.QueryRow("SELECT max(version) FROM _meta").Scan(&foundVersion); err != nil {
+		return xerrors.Errorf("invalid %s database version: no version found", name)
+	}
+
+	if foundVersion > schemaVersion {
+		return xerrors.Errorf("invalid %s database version: version %d is greater than the number of migrations %d", name, foundVersion, len(versionMigrations))
+	}
+
+	runVacuum := foundVersion != schemaVersion
+
+	for i := foundVersion + 1; i <= schemaVersion; i++ {
+		now := time.Now()
+
+		log.Infof("Migrating %s database to version %d...", name, i)
+
+		tx, err := db.BeginTx(ctx, nil)
+		if err != nil {
+			return xerrors.Errorf("failed to start %s database transaction: %w", name, err)
+		}
+		defer func() { _ = tx.Rollback() }()
```

Comment:
doing this in a loop is fine as only the tx that errors out will be rolled back when the function returns as all other txns will already have been committed and rollback is a no-op once a tx is committed.